src/LinearVesting.hs

```diff
@@ -1,19 +1,57 @@
-{-# OPTIONS_GHC -Wno-orphans #-}
-{-# OPTIONS_GHC -Wno-unused-imports #-}
-
-module LinearVesting (pvalidateVestingScriptValidator) where
-
-import Plutarch.Api.V1 (PAddress, PCredential (PPubKeyCredential, PScriptCredential))
-import Plutarch.Api.V2 (PPOSIXTime, PScriptContext, PScriptHash, PScriptPurpose (PSpending), PTxInInfo, PValidator)
-import Plutarch.DataRepr (PDataFields)
-import Plutarch.Extra.AssetClass (PAssetClassData, ptoScottEncoding)
+module LinearVesting (
+  PVestingDatum (..),
+  PVestingRedeemer (..),
+  VestingDatum (..),
+  VestingRedeemer (..),
+  pvalidateVestingPartialUnlock,
+  pvalidateVestingScriptValidator,
+) where
+
+import PlutusLedgerApi.V1.Value (AssetClass)
+import PlutusLedgerApi.V2 (Address)
+import PlutusTx qualified
+
+import Plutarch.Api.V1 (PAddress, PCredential (..))
+import Plutarch.Api.V2 (PCurrencySymbol, PScriptContext, PScriptHash, PScriptPurpose (PSpending), PTokenName, PTxInInfo, PValidator)
+import Plutarch.DataRepr (DerivePConstantViaData (..), PDataFields)
+import Plutarch.Extra.AssetClass (PAssetClass (..))
 import Plutarch.Extra.ScriptContext (pfindOutputsToAddress, pfindOwnInput, ptxSignedBy)
-import Plutarch.Extra.Time (PFullyBoundedTimeRange (PFullyBoundedTimeRange), passertFullyBoundedTimeRange)
 import Plutarch.Extra.Value (passetClassValueOf)
-import Plutarch.Positive (ptryPositive)
+import Plutarch.Lift (PConstantDecl, PUnsafeLiftDecl (..))
+import Plutarch.Monadic qualified as P
 import Plutarch.Prelude
-import Utils (pdivCeil, pgetLowerInclusiveTimeRange, pheadSingleton)
-import "plutarch-extra" Plutarch.Extra.TermCont (pguardC, pletC, pletFieldsC, pmatchC, ptraceC, ptryFromC)
+import "liqwid-plutarch-extra" Plutarch.Extra.TermCont (pguardC, pletC, pletFieldsC, pmatchC)
+
+import Conversions
+import Utils
+
+data VestingDatum = VestingDatum
+  { beneficiary :: Address
+  , vestingAsset :: AssetClass
+  , totalVestingQty :: Integer
+  , vestingPeriodStart :: Integer
+  , vestingPeriodEnd :: Integer
+  , firstUnlockPossibleAfter :: Integer
+  , totalInstallments :: Integer
+  }
+  deriving stock (Show)
+
+PlutusTx.makeLift ''VestingDatum
+PlutusTx.makeIsDataIndexed ''VestingDatum [('VestingDatum, 0)]
+
+data PAssetClassData (s :: S) = PAssetClassData (Term s (PDataRecord '["symbol" ':= PCurrencySymbol, "tokenName" ':= PTokenName]))
```

Comment:
Note: if the type with the same name is used from Plutarch.Extra, the datum deserialization fails with `unListData (Const 0 [...])`.

Reply:
no derivation in Plutarch.Extra.AssetClass?

Reply:
So, there are a couple, the one that we can use here is the data encoded one `PAssetClassData`. But that cannot be deserialized at runtime, due to the error I mentioned above.